controllers/openstackdataplanenode_controller.go

```diff
@@ -135,16 +141,59 @@ type Inventory struct {
 }
 
 // GenerateInventory yields a parsed Inventory
-func (r *OpenStackDataPlaneNodeReconciler) GenerateInventory(ctx context.Context, instance *dataplanev1beta1.OpenStackDataPlaneNode) error {
+func (r *OpenStackDataPlaneNodeReconciler) GenerateInventory(ctx context.Context, instance *dataplanev1beta1.OpenStackDataPlaneNode, instanceRole *dataplanev1beta1.OpenStackDataPlaneRole) error {
 	var err error
 
 	inventory := make(map[string]map[string]map[string]map[string]string)
 	all := make(map[string]map[string]map[string]string)
 	host := make(map[string]map[string]string)
 	hostVars := make(map[string]string)
-	hostVars["ansible_host"] = instance.Spec.Node.HostName
-	hostVars["ansible_user"] = instance.Spec.Node.AnsibleUser
-	hostVars["ansible_port"] = strconv.Itoa(instance.Spec.Node.AnsiblePort)
+
+	// Set values from the role template unless the instance already has them
+	if instance.Spec.Node.AnsibleUser != "" {
+		hostVars["ansible_user"] = instance.Spec.Node.AnsibleUser
+	} else {
+		hostVars["ansible_user"] = instanceRole.Spec.NodeTemplate.AnsibleUser
+	}
+	if instance.Spec.Node.AnsiblePort > 0 {
+		hostVars["ansible_port"] = strconv.Itoa(instance.Spec.Node.AnsiblePort)
+	} else {
+		hostVars["ansible_port"] = strconv.Itoa(instanceRole.Spec.NodeTemplate.AnsiblePort)
+	}
+	if instance.Spec.Node.Managed {
+		hostVars["managed"] = strconv.FormatBool(instance.Spec.Node.Managed)
+	} else {
+		hostVars["managed"] = strconv.FormatBool(instanceRole.Spec.NodeTemplate.Managed)
+	}
+	if instance.Spec.Node.ManagementNetwork != "" {
+		hostVars["management_network"] = instance.Spec.Node.ManagementNetwork
+	} else {
+		hostVars["management_network"] = instanceRole.Spec.NodeTemplate.ManagementNetwork
+	}
+	if instance.Spec.Node.NetworkConfig != instanceRole.Spec.NodeTemplate.NetworkConfig {
+		hostVars["network_config"] = fmt.Sprintf("{template: %s}",
+			instance.Spec.Node.NetworkConfig)
+	} else {
+		hostVars["network_config"] = fmt.Sprintf("{template: %s}",
+			instanceRole.Spec.NodeTemplate.NetworkConfig.Template)
+	}
```

Comment:
what about using getters here? Each "if" statement would be in a getter function:

`func (instance *dataplanev1beta1.OpenStackDataPlaneNode) GetAnsibleUser() string { ... }`

Reply:
As per our conversation, given the scope of this change it's ok having a getter like:

```
func (r *OpenStackDataPlaneNodeReconciler) GetAnsibleUser(instance *dataplanev1beta1.OpenStackDataPlaneNode, instanceRole *dataplanev1beta1.OpenStackDataPlaneRole) string {...}
```

that lives in the controller, where the two data structures can be processed.